zerver/views/video_calls.py

```diff
@@ -1,12 +1,129 @@
+        user_profile.save(update_fields=["zoom_token"])
+
+    return OAuth2Session(
+        settings.VIDEO_ZOOM_CLIENT_ID,
+        redirect_uri=user_profile.realm.uri + "/calls/zoom/complete",
+        scope=["meeting:write:admin"],
+        auto_refresh_url="https://zoom.us/oauth/token",
+        auto_refresh_kwargs={
+            "client_id": settings.VIDEO_ZOOM_CLIENT_ID,
+            "client_secret": settings.VIDEO_ZOOM_CLIENT_SECRET,
+        },
+        token_updater=token_updater,
+        token=user_profile.zoom_token,
+    )
+
+
+def get_zoom_oauth_state(request: HttpRequest) -> str:
+    # This is used to prevent CSRF attacks on the Zoom OAuth
+    # authentication flow.  We want this value to be unpredictable and
+    # tied to the session, but we don’t want to expose the main CSRF
+    # token directly to the Zoom server.
+
+    csrf.get_token(request)
+    return (
+        ""
+        if getattr(request, "_dont_enforce_csrf_checks", False)
+        else salted_hmac(
+            "Zulip Zoom OAuth state", request.META["CSRF_COOKIE"]
+        ).hexdigest()
+    )
 
+
+@zulip_login_required
+@never_cache
+def register_zoom_user(request: HttpRequest) -> HttpResponse:
+    oauth = zoom_oauth_session(request.user)
+    authorization_url, state = oauth.authorization_url(
+        "https://zoom.us/oauth/authorize", state=get_zoom_oauth_state(request)
+    )
+    return redirect(authorization_url)
+
+
+@zulip_login_required
+@never_cache
 @has_request_variables
-def get_zoom_url(request: HttpRequest, user_profile: UserProfile) -> HttpResponse:
-    return json_success({'zoom_url': get_zoom_video_call_url(
-        user_profile.realm
-    )})
+def complete_zoom_user(
+    request: HttpRequest, code: str = REQ(), state: str = REQ()
+) -> HttpResponse:
+    if not constant_time_compare(state, get_zoom_oauth_state(request)):
+        raise JsonableError(_("Invalid Zoom OAuth state"))
+
+    oauth = zoom_oauth_session(request.user)
+    try:
+        request.user.zoom_token = oauth.fetch_token(
+            "https://zoom.us/oauth/token",
+            code=code,
+            client_secret=settings.VIDEO_ZOOM_CLIENT_SECRET,
+        )
+    except OAuth2Error:
+        raise JsonableError(_("Invalid Zoom credentials"))
+
+    request.user.save(update_fields=["zoom_token"])
+    send_event(
+        request.user.realm, dict(type="has_zoom_token", value=True), [request.user.id]
+    )
+    return render(request, "zerver/close_window.html")
+
+
+def make_zoom_video_call(
+    request: HttpRequest, user_profile: UserProfile
+) -> HttpResponse:
+    oauth = zoom_oauth_session(user_profile)
+    try:
+        res = oauth.post("https://api.zoom.us/v2/users/me/meetings", json={})
+    except OAuth2Error:
+        request.user.zoom_token = None
+        request.user.save(update_fields=["zoom_token"])
+        send_event(
+            request.user.realm,
+            dict(type="has_zoom_token", value=False),
+            [request.user.id],
+        )
```

Comment:
Please move the `send_event` calls to `zerver.lib.actions.py`; as a codebase policy, we don't send events from views.